### PR TITLE
Go: Add Improper LDAP Authentication query (CWE-287)

### PR DESCRIPTION
**Pull request by maikypedia**
This pull request adds a query for Improper LDAP Authentication to prevent attackers use an empty password. I am not very familiar with CodeQL Go, and I have been struggling to generate correct expected files 🙃.

Looking forward to your suggestions.

### CODE REVIEW

**Comment by mbg** [comment]
Hi @maikypedia, thank you for opening this pull request!

Looking at [the documentation for `Bind`](https://pkg.go.dev/github.com/go-ldap/ldap/v3#Conn.Bind), it seems that an empty password is not accepted by that method. Would it be correct in saying that this vulnerability cannot be exploited if that is the case?

**Comment by maikypedia** [comment]
> Hi @maikypedia, thank you for opening this pull request!
> 
> Looking at [the documentation for `Bind`](https://pkg.go.dev/github.com/go-ldap/ldap/v3#Conn.Bind), it seems that an empty password is not accepted by that method. Would it be correct in saying that this vulnerability cannot be exploited if that is the case?

Hi @mbg , I didn't realize that in v3 empty password option with Bind is disabled, sorry for the inconvenience 😅

**Comment by maikypedia** [comment]
Would it be worth adding `Bind` as sink for v2 and `UnauthenticatedBind` for v3? As a warning query.

**Comment by maikypedia** [comment]
ping @mbg

**Comment by mbg** [comment]
Hi @maikypedia, sorry, I must have missed your question. Thank you for pinging me to bring this to my attention.

For `Bind` in v2, the query does make sense so it could be worth adding and I would be happy to review that.

For `UnauthenticatedBind`, I am not sure it makes sense to add a query to detect its use. The name of the method already notes that it is unauthenticated so using it would have to be deliberate and intentional.

File: go/ql/test/experimental/CWE-287/ImproperLdapAuth.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+	if err != nil {
+		log.Fatalf("LDAP bind failed: %v", err)
+	}
+}
+
+func main() {
+	bad(nil, nil)
+	good1(nil, nil)
+	good2(nil, nil)
+	bad2(nil, nil)
```
Comment (owen-mc) [inline go/ql/test/experimental/CWE-287/ImproperLdapAuth.go:98]:
```suggestion
	bad2(nil)
```

File: go/ql/test/experimental/CWE-287/ImproperLdapAuth.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"regexp"
+
+	ldap "gopkg.in/ldap.v2"
+)
+
+func bad(w http.ResponseWriter, req *http.Request) (interface{}, error) {
```
Comment (owen-mc) [inline go/ql/test/experimental/CWE-287/ImproperLdapAuth.go:12]: This function is missing a return statement. And the next two functions.

File: go/ql/src/experimental/CWE-287/ImproperLdapAuth.qll [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+ * A vulnerable argument to `go-ldap` or `ldap`'s `bind` function (Only v2).
+ */
+private class GoLdapBindSink extends LdapAuthSink {
+  GoLdapBindSink() {
+    exists(Method meth, string base, string t, string m |
+      t = ["Conn"] and
+      meth.hasQualifiedName(["gopkg.in/ldap.v2"], t, m) and
+      this = meth.getACall().getArgument(1)
+    |
+      base = ["Bind"] and m = base
```
Comment (owen-mc) [inline go/ql/src/experimental/CWE-287/ImproperLdapAuth.qll:30]:
```suggestion
    exists(Method meth |
      meth.hasQualifiedName("gopkg.in/ldap.v2", "Conn", "Bind") and
      this = meth.getACall().getArgument(1)
```

**Review by owen-mc — CHANGES_REQUESTED** [review]
Your test isn't giving any results - it should be finding the bad cases. You may need to make sure that the test file builds correctly (try running `go build` in the directory that it is in to see what the errors are).

**Comment by owen-mc** [comment]
> The following files need to be reformatted using gofmt or have compilation errors:
> Error: ./ql/src/experimental/CWE-287/examples/LdapAuthGood.go:1:1: expected 'package', found 'func'
> Error: make: *** [Makefile:35: check-formatting] Error 1
> Error: ./ql/src/experimental/CWE-287/examples/LdapAuthBad.go:1:1: expected 'package', found 'func'

**Review by owen-mc — CHANGES_REQUESTED** [review]
Well done for generating stubs. You've actually put them in two places by mistake - you can remove all changes to `go/vendor` and keep the ones in the test folder. Your tests now generate the right results now, but they will also generate some build errors because of missing return statements. Please address that and the rest of my previous review comments.

**Comment by maikypedia** [comment]
I don't know if I've forgotten something, I think that's all for now. 😗

**Review by owen-mc — CHANGES_REQUESTED** [review]
Please don't make any changes to `go/vendor/modules.txt`. Please add a package declaration and imports to make `LdapAuthBad.go` and `LdapAuthGood.go` valid go files.

**Comment by maikypedia** [comment]
Hi @owen-mc ! Sorry for the delay, I had some exams these weeks 😅 In the example files of other experimental queries they do not include the imports of libraries that are not builtin, should I leave it like that?

**Comment by owen-mc** [comment]
@maikypedia The reason for making these valid go files is so that `gofmt` can parse them, because it can highlight some syntax errors. It seems to have passed. I see now that imports aren't actually needed, just the package directive. I would leave it as it is.

Your test is failing. It couldn't build the go file. I think there is a missing `go.mod` file.

**Comment by maikypedia** [comment]
> @maikypedia The reason for making these valid go files is so that `gofmt` can parse them, because it can highlight some syntax errors. It seems to have passed. I see now that imports aren't actually needed, just the package directive. I would leave it as it is.
> 
> Your test is failing. It couldn't build the go file. I think there is a missing `go.mod` file.

Done :), thanks

File: go/ql/src/experimental/CWE-287/ImproperLdapAuth.qll [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  ImproperLdapAuthConfiguration() { this = "Improper LDAP Auth" }
+
+  override predicate isSource(DataFlow::Node source) {
+    source instanceof UntrustedFlowSource or source instanceof EmptyString
+  }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof LdapAuthSink }
+
+  override predicate isSanitizer(DataFlow::Node node) { node instanceof LdapSanitizer }
+}
```
Comment (owen-mc) [inline go/ql/src/experimental/CWE-287/ImproperLdapAuth.qll:83]: `TaintTracking::Configuration` is deprecated. Please convert this to use the new API for the dataflow library. See [this PR](https://github.com/github/codeql/pull/13820) for lots of examples. It should be a straightforward change, so let me know if you have any problems. The main thing that might trip you up is that you have to use `DataFlow::ConfigSig` and then `TaintTracking::Global`. (And no need to leave a deprecated class extending `TaintTracking::Configuration` around.)

**Review by owen-mc — CHANGES_REQUESTED** [review]

**Comment by maikypedia** [comment]
Done 😃 👍

**Review by owen-mc — APPROVED** [review]

**Pull request merged**